### PR TITLE
Game refactor

### PR DESCRIPTION
**Pull request by jamogriff**
Large changes:
- Refactor individual ship instances into a nested hash data structure.
- Tests added for griffith and merali algorithms

Other changes:
- Refactor Game class and add new tests
- Fixes vertical/horizontal failing tests in board_spec

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -1,59 +1,66 @@
...
   end
 
 
-  # Validates user's coordinates for ship placement
-  def get_user_coordinates(ship)
+  # Validates player's coordinates for ship placement
+  def get_player_coordinates(ship)
     correct = false
 
     while correct == false
```
Comment (aliyamerali) [inline lib/game.rb]: Do we need to do a nested while + if statement here? it looks like we're only iterating over valid response in this one (not checking for already-shot-shot) so I think aa single while loop could do it, right? 

```ruby
response = gets.chomp
while @player_board.valid_placement?(ship, response) == false
    puts "Those are invalid coordinates. Please try again"
    response = gets.chomp
end
```

Comment (aliyamerali) [reply lib/game.rb:66]: Or are we just missing the piece here that checks if the shot has been fired already? I think this is where we had put it in but I don't see the check for it?

Comment (jamogriff) [reply lib/game.rb]: Good refactor! Implemented!

Comment (jamogriff) [reply lib/game.rb]: See here: c1c0d4868b13acf790810b76d835a421861e4409

**Pull request merged**